### DISCUSSION TITLE
Deformation field show nothing in the plot

### DISCUSSION THREAD

**Discussion by YuliangXiaoYLX** [Project-MONAI/MONAI#4504] [Q&A] [answered]
# Issue
Hi, I am working on the deep atlas with 128x128x96 CT scans which are normalized to [0,1]. When I used the trained registration network to make inference, the deformation field plot showed nothing. I checked the value of deformation field and its range is between [-1,1]. Also, I used the deformation field learned from the registration network to warped the moving segmentation, the result seems only to have translation which are shown below (green line shows only translation and small rotation). Do you have any idea about this issue? Thanks in advance
# Result
### deformation field plot
![reg_net_infer_right_IR_184_to_right_IR_189](https://user-images.githubusercontent.com/70731556/173703396-b4436bfd-831e-420d-83c1-c048fea1ba2d.png)
### Comparison between warped segmentation, moving segmentation and target segmentation
![Screen Shot 2022-06-14 at 19 05 27](https://user-images.githubusercontent.com/70731556/173704023-d94d86ff-5efc-4653-9c10-e71b3f821c45.png)

**Comment by ebrahimebrahim** [comment]
> When I used the trained registration network to make inference, the deformation field plot showed nothing

Maybe regularization is set too high?

> Also, I used the deformation field learned from the registration network to warped the moving segmentation, the result seems only to have translation which are shown below

Often registration is done in two steps: first an affine or euclidean transform, then a "deformable registration" map.
The type of registration model produced by the DeepAtlas tutorial notebook is shown to work well as the deformable registration step; i.e. doing affine registration separately beforehand may be necessary to get good results.

I don't know why you're only getting this translation and small rotation. Maybe you can share how you are warping segmentations.

**Reply by YuliangXiaoYLX** [reply]
@ebrahimebrahim , thanks for your reply. The regularization lambda for registration is ```lambda_r = 7.5 * (image_scale / 96)**2``` and ```image_scale``` is 96. The code I used to warp segmentation and scan is shown below:
```
reg_net.load_state_dict(torch.load(best_model_path))
reg_net.to(device)
reg_net.eval()
img12 = data_item['img12'].unsqueeze(0).to(device)
gt_raw_seg = data_item['seg1'].unsqueeze(0).to(device)
moving_raw_seg = data_item['seg2'].unsqueeze(0).to(device)
moving_seg = monai.networks.one_hot(moving_raw_seg, num_labels)
gt_seg = monai.networks.one_hot(gt_raw_seg, num_labels)
with torch.no_grad():
      reg_net_example_output = reg_net(img12)
  
example_warped_image = warp(
    img12[:, [1], :, :, :],  # moving image
    reg_net_example_output  # warping
)
example_warped_seg = warp_nearest(
    moving_seg,       # moving seg
    reg_net_example_output # warping
)
```
``` data_item``` is the data_pairs loading with CacheDatset and containing ['img12', 'seg1', 'seg2']

**Reply by YuliangXiaoYLX** [reply]
> > When I used the trained registration network to make inference, the deformation field plot showed nothing
> 
> Maybe regularization is set too high?
> 
> > Also, I used the deformation field learned from the registration network to warped the moving segmentation, the result seems only to have translation which are shown below
> 
> Often registration is done in two steps: first an affine or euclidean transform, then a "deformable registration" map. The type of registration model produced by the DeepAtlas tutorial notebook is shown to work well as the deformable registration step; i.e. doing affine registration separately beforehand may be necessary to get good results.
> 
> I don't know why you're only getting this translation and small rotation. Maybe you can share how you are warping segmentations.

Do you have any idea why plot of deformation field shows nothing?

**Reply by ebrahimebrahim** [reply]
> Do you have any idea why plot of deformation field shows nothing?

You could try lowering the regularization and see if you get anything. It can be a lot of trouble to fine-tune the regularization for your dataset. The `(image_scale / 96)**2` part is there to help you get in the right ballpark if you have a different image scale. Try for example a `lambda_r` that is one tenth that size and see if you get interesting looking deformation fields, etc.

Also: _should_ the deformations be larger? Do you expect large deformations for this dataset?

As for the segmentation getting translated like that when the deformation fields are close to zero, I don't know why that is happening. Is that segmentation a very small part of the overall volume? Then I might worry about a misaligned resampling process, and I'd want to look into that.

**Reply by YuliangXiaoYLX** [reply]
> > Do you have any idea why plot of deformation field shows nothing?
> 
> You could try lowering the regularization and see if you get anything. It can be a lot of trouble to fine-tune the regularization for your dataset. The `(image_scale / 96)**2` part is there to help you get in the right ballpark if you have a different image scale. Try for example a `lambda_r` that is one tenth that size and see if you get interesting looking deformation fields, etc.
> 
> Also: _should_ the deformations be larger? Do you expect large deformations for this dataset?
> 
> As for the segmentation getting translated like that when the deformation fields are close to zero, I don't know why that is happening. Is that segmentation a very small part of the overall volume? Then I might worry about a misaligned resampling process, and I'd want to look into that.

I made test just now on the registration network. I only forward one step and then use the output (displacement field) of the network to warp the moving image. Good news was I got something and it was not bad. Here are some results:
![Screen Shot 2022-06-14 at 22 24 37](https://user-images.githubusercontent.com/70731556/173725476-a46342c7-08f8-4f2c-80c2-56bdc0dd5c52.png)

![Screen Shot 2022-06-14 at 22 24 45](https://user-images.githubusercontent.com/70731556/173725486-09aeb415-c2d9-41c3-9947-123f8e71333f.png)

For the segmentation, it takes some part of the volume instead a very small one. The dataset has raw resolution about 512x512x256. Since the GPU has insufficient memory, I extracted ROI which contains the label and pad them to 128x128x96. Also, the scans and segmentation are all in the nii format, they have same coordinate, origin and spacing.

**Reply by YuliangXiaoYLX** [reply]
> > Do you have any idea why plot of deformation field shows nothing?
> 
> You could try lowering the regularization and see if you get anything. It can be a lot of trouble to fine-tune the regularization for your dataset. The `(image_scale / 96)**2` part is there to help you get in the right ballpark if you have a different image scale. Try for example a `lambda_r` that is one tenth that size and see if you get interesting looking deformation fields, etc.
> 
> Also: _should_ the deformations be larger? Do you expect large deformations for this dataset?
> 
> As for the segmentation getting translated like that when the deformation fields are close to zero, I don't know why that is happening. Is that segmentation a very small part of the overall volume? Then I might worry about a misaligned resampling process, and I'd want to look into that.

Maybe I should have a try to lower the regularization and see the result, thanks for your suggestion and I will be back if I get some results.

**Reply by ebrahimebrahim** [reply]
> Maybe I should have a try to lower the regularization and see the result, thanks for your suggestion and I will be back if I get some results.

By the way I noticed that in your `lambda_r` you changed `(image_scale / 128)**2` to `(image_scale / 96)**2`. I think you want to leave the denominator as 128. The idea is that if you take `image_scale` down to 96, then `lambda_r` should be scaled down accordingly. This needs to happen because the bending energy loss depends on the image scale. I added an option for a scale-independent mode, but I think it is not yet released in MONAI.

> I extracted ROI which contains the label and pad them to 128x128x96

I wonder also if there's an issue with resizing, since the deepatlas set up to work with all dimensions being the same size. Maybe the weird translation you are seeing is coming from an error in applying a transform meant for 96x96x96 images to a 128x128x96 image

**Reply by YuliangXiaoYLX** [reply]
> > Maybe I should have a try to lower the regularization and see the result, thanks for your suggestion and I will be back if I get some results.
> 
> By the way I noticed that in your `lambda_r` you changed `(image_scale / 128)**2` to `(image_scale / 96)**2`. I think you want to leave the denominator as 128. The idea is that if you take `image_scale` down to 96, then `lambda_r` should be scaled down accordingly. This needs to happen because the bending energy loss depends on the image scale. I added an option for a scale-independent mode, but I think it is not yet released in MONAI.
> 
> > I extracted ROI which contains the label and pad them to 128x128x96
> 
> I wonder also if there's an issue with resizing, since the deepatlas set up to work with all dimensions being the same size. Maybe the weird translation you are seeing is coming from an error in applying a transform meant for 96x96x96 images to a 128x128x96 image

The regularization makes sense to me. But for the dataset dimension, I tried on VoxelMorph which uses same architecture that deep atlas uses. The VoxelMorph works well with 128x128x96. I thought deep atlas should be fine with same dimension. Also, the deepatlas paper also uses the unequal size which is shown below:
![Screen Shot 2022-06-14 at 23 24 06](https://user-images.githubusercontent.com/70731556/173730577-8f877492-69b1-489c-baba-bb493ced2285.png)

**Reply by ebrahimebrahim** [reply]
I suppose any dimensions should be fine as long they work with the strided convolutions and as long as you are consistent about the coordinate system in which your deformation field is defined.

**Reply by wyli** [reply]
(btw BendingEnergyLoss with a `normalize` option is available in monai v0.9 released earlier this week https://docs.monai.io/en/stable/losses.html#monai.losses.BendingEnergyLoss)

**Reply by YuliangXiaoYLX** [reply]
@wyli Thank you so much, I will take a look

**Reply by YuliangXiaoYLX** [reply]
> (btw BendingEnergyLoss with a `normalize` option is available in monai v0.9 released earlier this week https://docs.monai.io/en/stable/losses.html#monai.losses.BendingEnergyLoss)

@wyli @ebrahimebrahim  I tried on normalized BendingEnergyLoss for regularization. Since, it is now  roughly invariant to the image scale, the lambda of regularization is 0.0005. But I still cannot get the vector field. Do you have any ideas?

**Reply by ebrahimebrahim** [reply]
Do you get a zero deformation field no matter how small `lambda_r` is? e.g. even if you turn off regularization completely?

**Reply by YuliangXiaoYLX** [reply]
> Do you get a zero deformation field no matter how small `lambda_r` is? e.g. even if you turn off regularization completely?

Same as before, I did not get zero deformation field but the plot shows nothing or the value is small within [-1,1] to be seen on the plot

**Reply by YuliangXiaoYLX** [reply]
> Do you get a zero deformation field no matter how small `lambda_r` is? e.g. even if you turn off regularization completely?

I tried two ways. First, use the normalized BendingEnergyLoss, second is using non-normalized but lambda of regularization is ```7.5 * (image scale / 128) ** 2``` which is 7.5 for my dataset. Both of them give me similar results

**Reply by ebrahimebrahim** [reply]
Normalized is easier to use because you don't need the $(\text{scale})^2$ term. I still wonder, do you get a close-to-zero deformation field no matter how small `lambda_r` is? Even when `lambda_r` is set to zero? If that is happening and your image pairs aren't similar to begin with, then I am puzzled. In that case, feel free to share more about what you are doing.

**Reply by ebrahimebrahim** [reply]
By the way to help convert between normalized bending energy loss and non-normalized, the relation is
$$\text{(normalized BEL)} = \text{(non-normalized BEL)} \cdot s^2$$
for an $s\times s\times s$ image. The relation is more complicated if the image dimensions are not all equal, but this relation can still be used to get in the right ballpark.

**Reply by YuliangXiaoYLX** [reply]
> Normalized is easier to use because you don't need the (scale)2 term. I still wonder, do you get a close-to-zero deformation field no matter how small `lambda_r` is? Even when `lambda_r` is set to zero? If that is happening and your image pairs aren't similar to begin with, then I am puzzled. In that case, feel free to share more about what you are doing.

I have not tried on zero ```lambda_r```, and the issue is I can get some rational deformation field one step forward but why did I get such small value after some epochs? In fact, my images are a little bit similar to each other, I can share some scans to you if you wanna know more about my issue. 
[left_IR_135.nii.gz](https://github.com/Project-MONAI/MONAI/files/8915154/left_IR_135.nii.gz)
[left_IR_138.nii.gz](https://github.com/Project-MONAI/MONAI/files/8915155/left_IR_138.nii.gz)
[left_IR_139.nii.gz](https://github.com/Project-MONAI/MONAI/files/8915156/left_IR_139.nii.gz)
[right_IR_135.nii.gz](https://github.com/Project-MONAI/MONAI/files/8915157/right_IR_135.nii.gz)
[right_IR_138.nii.gz](https://github.com/Project-MONAI/MONAI/files/8915158/right_IR_138.nii.gz)
[right_IR_139.nii.gz](https://github.com/Project-MONAI/MONAI/files/8915159/right_IR_139.nii.gz)

**Reply by ebrahimebrahim** [reply]
Try smaller and smaller `lambda_r` until the deformation fields become reasonable to you after training converges

**Reply by YuliangXiaoYLX** [reply]
> By the way to help convert between normalized bending energy loss and non-normalized, the relation is (normalized BEL)=(non-normalized BEL)⋅s2 for an s×s×s image. The relation is more complicated if the image dimensions are not all equal, but this relation can still be used to get in the right ballpark.

I tried to use normalized Bending Loss and it still give unexpected results, the ```lambda_r = 0.0005``` for this case

**Reply by YuliangXiaoYLX** [reply]
> Try smaller and smaller `lambda_r` until the deformation fields become reasonable to you after training converges

You mean try smaller ```lambda_r``` for the normalized Bending Loss or non-normalized one?

**Reply by ebrahimebrahim** [reply]
Either. For example if you prefer normalized, where you have tried `lambda_r = 5e-4`, try also `lambda_r = 5e-5`, then `lambda_r = 5e-6`, etc.

**Reply by YuliangXiaoYLX** [reply]
> Either. For example if you prefer normalized, where you have tried `lambda_r = 5e-4`, try also `lambda_r = 5e-5`, then `lambda_r = 5e-6`, etc.

Except decreasing the ```lambda_r```, are there anything that I can improve to get rational displacement field?

**Reply by ebrahimebrahim** [reply, marked as answer]
You can try a different image similarity loss function. For example maybe a simple mean squared error loss can work for you, if maybe LNCC is causing problems for training. You can try changing how much you focus on the anatomy loss (warped segmentation dice score) versus image similarity loss. Maybe in the case of your dataset, one of those is much more helpful than the other, or maybe you will discover an error in one of them that is causing it to break training.

**Reply by YuliangXiaoYLX** [reply]
> You can try a different image similarity loss function. For example maybe a simple mean squared error loss can work for you, if maybe LNCC is causing problems for training. You can try changing how much you focus on the anatomy loss (warped segmentation dice score) versus image similarity loss. Maybe in the case of your dataset, one of those is much more helpful than the other, or maybe you will discover an error in one of them that is causing it to break training.

Gotch that. Yeap, ```MSE``` is also used in ```VoxelMorph```. I also trained the same dataset with ```VoxelMorph```, the deformation field was much better than the ```deep atlas``` and I used ```MSE```for similarity loss, but regularization is also different from what I used in the ```deep atlas```
